src/wasm-interpreter.h

```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
     : ModuleInstanceBase(wasm, externalInterface) {}
 };
 
+// Evaluates an expression given its surrounding context. Errors if we hit
+// anything that can't be evaluated down to a constant.
+class ContextAwareExpressionRunner final
+  : public ExpressionRunner<ContextAwareExpressionRunner> {
+
+public:
+  // Map of `local.get`s to their respective values.
+  typedef std::unordered_map<LocalGet*, Literals> GetValues;
+
+  // Whether we are trying to precompute down to an expression (which we can
+  // do on say 5 + 6) or to a value (which we can't do on a local.tee that
+  // flows a 7 through it). When we want to replace the expression, we can
+  // only do so when it has no side effects. When we don't care about
+  // replacing the expression, we just want to know if it will contain a known
+  // constant.
+  enum FlagValues {
+    // Just evaluate the expression by default, where we can ignore some side
+    // effects like those of a `local.tee`, but not others like traps.
+    DEFAULT = 0,
+    // We are going to replace the expression afterwards, so side effects
+    // including those of `local.tee`s for example must be retained.
+    REPLACE = 1 << 0,
+    // We are going to execute the runner in a function-parallel scenario, so we
+    // cannot perform traversal of nodes that might become modified
+    // non-deterministically, like function calls, where the called function
+    // might or might not have been optimized already to something we can
+    // traverse successfully.
+    PARALLEL = 1 << 1
+  };
+  // Flags indicating special requirements, for example whether we are just
+  // evaluating (default), also going to replace the expression afterwards or
+  // executing in a function-parallel scenario. See FlagValues.
+  typedef uint32_t Flags;
+
+  // Special break target indicating a flow not evaluating to a constant
+  static const Name NONCONSTANT_FLOW;
+
+  // Special exception indicating a flow not evaluating to a constant
+  struct NonconstantException {
+  }; // TODO: use a flow with a special name, as this is likely very slow
+
+  ContextAwareExpressionRunner(Module* module,
+                               Flags flags,
+                               Index maxDepth,
+                               GetValues* getValues = nullptr);
+
+  // Gets the module this runner belongs to.
+  Module* getModule();
+
+  // Sets a known local value to use. Returns `true` if the value is actually
+  // constant.
```

Comment:
"actually constant" seems a little inaccurate. how about "is concrete (i.e., not unreachable)"